File: src/bim_review_agent/application/agent/registry.py
"""Typed, policy-filtered tool registration and dispatch."""

from __future__ import annotations

from collections.abc import Callable, Iterable
from dataclasses import dataclass
from typing import Any

from pydantic import BaseModel, ValidationError

from bim_review_agent.application.agent.schemas import AgentStopReason, ToolDescriptor, ToolEffect

ToolHandler = Callable[[BaseModel, Any], BaseModel | dict[str, Any]]


class ToolDispatchError(RuntimeError):
    def __init__(self, reason: AgentStopReason, message: str) -> None:
        super().__init__(message)
        self.reason = reason


@dataclass(frozen=True, slots=True)
class RegisteredTool:
    descriptor: ToolDescriptor
    input_model: type[BaseModel]
    output_model: type[BaseModel]
    handler: ToolHandler


class ToolRegistry:
    """Own tool schemas and handlers; expose only an Agent's allowlisted catalogue."""

    def __init__(self) -> None:
        self._tools: dict[str, RegisteredTool] = {}

    def register(
        self,
        *,
        name: str,
        version: str,
        description: str,
        effect: ToolEffect,
        input_model: type[BaseModel],
        output_model: type[BaseModel],
        handler: ToolHandler,
    ) -> None:
        if name in self._tools:
            raise ValueError(f"Tool already registered: {name}")
        descriptor = ToolDescriptor(
            name=name,
            version=version,
            description=description,
            effect=effect,
            input_schema=input_model.model_json_schema(),
            output_schema=output_model.model_json_schema(),
        )
        self._tools[name] = RegisteredTool(
            descriptor=descriptor,
            input_model=input_model,
            output_model=output_model,
            handler=handler,
        )

    def contains(self, name: str) -> bool:
        return name in self._tools

    def catalogue(self, allowed_tools: Iterable[str]) -> tuple[ToolDescriptor, ...]:
        allowed = set(allowed_tools)
        return tuple(
            registered.descriptor
            for name, registered in sorted(self._tools.items())
            if name in allowed
        )

    def execute(self, name: str, arguments: dict[str, Any], context: Any) -> dict[str, Any]:
        registered = self._tools.get(name)
        if registered is None:
            raise ToolDispatchError(
                AgentStopReason.TOOL_NOT_FOUND,
                f"Requested tool is not registered: {name}",
            )
        try:
            validated_input = registered.input_model.model_validate(arguments)
        except ValidationError as exc:
            raise ToolDispatchError(
                AgentStopReason.TOOL_INPUT_INVALID,
                f"Input for tool {name} does not match its declared schema.",
            ) from exc

        try:
            raw_output = registered.handler(validated_input, context)
        except ToolDispatchError:
            raise
        except Exception as exc:
            raise ToolDispatchError(
                AgentStopReason.TOOL_EXECUTION_FAILED,
                f"Tool {name} failed during execution.",
            ) from exc

        try:
            validated_output = registered.output_model.model_validate(raw_output)
        except ValidationError as exc:
            raise ToolDispatchError(
                AgentStopReason.TOOL_OUTPUT_INVALID,
                f"Output from tool {name} does not match its declared schema.",
            ) from exc
        return validated_output.model_dump(mode="json")

File: apps/gpt-sites/src/runtime/agent/registry.ts
import type {
  AgentStopReason,
  JsonObject,
  ToolDescriptor,
} from "../../contracts/agent";
import { requireJsonObject } from "./actions";

export type ToolValidator = (value: unknown) => JsonObject;
export type ToolHandler<Context> = (
  input: JsonObject,
  context: Context,
) => JsonObject | Promise<JsonObject>;

export type RegisteredTool<Context> = {
  descriptor: ToolDescriptor;
  validate_input: ToolValidator;
  validate_output: ToolValidator;
  handler: ToolHandler<Context>;
  canonical_review_output_key?: string;
  public_observation?: (output: JsonObject) => JsonObject;
};

export type ToolExecution = {
  output: JsonObject;
  public_output: JsonObject;
  canonical_review_run_id: string | null;
};

export class ToolDispatchError extends Error {
  constructor(
    readonly reason: AgentStopReason,
    message: string,
  ) {
    super(message);
    this.name = "ToolDispatchError";
  }
}

export class ToolRegistry<Context> {
  readonly #tools = new Map<string, RegisteredTool<Context>>();

  register(tool: RegisteredTool<Context>): void {
    if (this.#tools.has(tool.descriptor.name)) {
      throw new Error(`Tool already registered: ${tool.descriptor.name}`);
    }
    this.#tools.set(tool.descriptor.name, tool);
  }

  contains(name: string): boolean {
    return this.#tools.has(name);
  }

  isCanonicalReviewTool(name: string): boolean {
    return Boolean(this.#tools.get(name)?.canonical_review_output_key);
  }

  catalogue(allowedTools: readonly string[]): ToolDescriptor[] {
    const allowed = new Set(allowedTools);
    return Array.from(this.#tools.values())
      .filter((tool) => allowed.has(tool.descriptor.name))
      .sort((left, right) =>
        left.descriptor.name.localeCompare(right.descriptor.name, "en"),
      )
      .map((tool) => structuredClone(tool.descriptor));
  }

  async execute(
    name: string,
    argumentsValue: JsonObject,
    context: Context,
  ): Promise<ToolExecution> {
    const tool = this.#tools.get(name);
    if (!tool) {
      throw new ToolDispatchError(
        "TOOL_NOT_FOUND",
        `Requested tool is not registered: ${name}`,
      );
    }

    let input: JsonObject;
    try {
      input = tool.validate_input(argumentsValue);
    } catch {
      throw new ToolDispatchError(
        "TOOL_INPUT_INVALID",
        `Input for tool ${name} does not match its declared schema.`,
      );
    }

    let rawOutput: unknown;
    try {
      rawOutput = await tool.handler(input, context);
    } catch (error) {
      if (error instanceof ToolDispatchError) throw error;
      throw new ToolDispatchError(
        "TOOL_EXECUTION_FAILED",
        `Tool ${name} failed during execution.`,
      );
    }

    let output: JsonObject;
    try {
      output = requireJsonObject(tool.validate_output(rawOutput));
    } catch {
      throw new ToolDispatchError(
        "TOOL_OUTPUT_INVALID",
        `Output from tool ${name} does not match its declared schema.`,
      );
    }

    const outputKey = tool.canonical_review_output_key;
    const candidate = outputKey ? output[outputKey] : null;
    if (outputKey && (typeof candidate !== "string" || !candidate)) {
      throw new ToolDispatchError(
        "TOOL_OUTPUT_INVALID",
        `Output from tool ${name} did not identify its canonical ReviewRun.`,
      );
    }
    let publicOutput: JsonObject;
    try {
      publicOutput = requireJsonObject(
        tool.public_observation ? tool.public_observation(output) : output,
      );
    } catch {
      throw new ToolDispatchError(
        "TOOL_OUTPUT_INVALID",
        `Public observation from tool ${name} does not match its bounded contract.`,
      );
    }
    return {
      output,
      public_output: publicOutput,
      canonical_review_run_id: typeof candidate === "string" ? candidate : null,
    };
  }
}
